landoapi/transplants.py

```diff
@@ -495,43 +606,330 @@ def check_landing_blockers(
+            )
+
+            if reason is not None:
+                assessment.warnings.append(reason)
+
+    return assessment
+
+
+def assess_transplant_request(
+    phab: PhabricatorClient,
+    supported_repos: dict[str, Repo],
     stack_data: RevisionData,
+    stack: RevisionStack,
+    relman_group_phid: str,
     data_policy_review_phid: str,
-):
-    """Build all transplant blocker checks that need extra Phabricator data"""
-    assert all((isinstance(r, Repo) for r in repositories.values()))
-
-    return DEFAULT_OTHER_BLOCKER_CHECKS + [
-        # Configure uplift check with extra data.
-        check_uplift_approval(relman_group_phid, repositories, stack_data),
-        # Configure data policy check with the project PHID.
-        check_revision_data_classification(data_policy_review_phid),
-    ]
+    landing_assessment: Optional[LandingAssessmentState] = None,
+) -> tuple[TransplantAssessment, TransplantAssessmentState]:
+    """Assess the transplant request."""
+    # Get data from Lando.
+    landable_repos = get_landable_repos_for_revision_data(stack_data, supported_repos)
+
+    involved_phids = set()
+    reviewers = {}
+    for revision in stack_data.revisions.values():
+        involved_phids.update(gather_involved_phids(revision))
+        reviewers[revision["phid"]] = get_collated_reviewers(revision)
+
+    # Get more Phabricator data.
+    involved_phids = list(involved_phids)
+    users = user_search(phab, involved_phids)
+    projects = project_search(phab, involved_phids)
+
+    secure_project_phid = get_secure_project_phid(phab)
+    testing_tag_project_phids = get_testing_tag_project_phids(phab)
+    testing_policy_phid = get_testing_policy_phid(phab)
+
+    # Create a copy of the stack where we will remove revisions that are blocked from
+    # landing, leaving a graph where each path is landable.
+    landable_stack = copy.deepcopy(stack)
+
+    # We only want to consider paths starting from the open revisions
+    # do grab the status for all revisions.
+    statuses = {
+        phid: PhabricatorRevisionStatus.from_status(
+            PhabricatorClient.expect(revision, "fields", "status", "value")
+        )
+        for phid, revision in stack_data.revisions.items()
+    }
+
+    transplant_state = TransplantAssessmentState(
+        phab=phab,
+        stack_data=stack_data,
+        stack=stack,
+        statuses=statuses,
+        landable_stack=landable_stack,
+        landable_repos=landable_repos,
+        supported_repos=supported_repos,
+        reviewers=reviewers,
+        users=users,
+        projects=projects,
+        data_policy_review_phid=data_policy_review_phid,
+        relman_group_phid=relman_group_phid,
+        secure_project_phid=secure_project_phid,
+        testing_tag_project_phids=testing_tag_project_phids,
+        testing_policy_phid=testing_policy_phid,
+        landing_assessment=landing_assessment,
```

Comment:
Too many arguments are being passed here that we could be fetching or calculating elsewhere. This should at most be:

```suggestion
        stack_data=stack_data,
        stack=stack,
        statuses=statuses,
        landable_stack=landable_stack,
        reviewers=reviewers,
        users=users,
        projects=projects,
        landing_assessment=landing_assessment,
```
For example, `landable_repos` isn't used in this method, even though it is calculated from two different variables, both which can be used in the definition of `TransplantAssessmentState`.

Reply:
For the most part, I've been trying to make as few changes to the existing checks as possible. `landable_repos` is used by multiple checks, so I've carried it forward as something that is contained in the `TransplantAssessmentState` so both checks have access to it after it is built once. I'm going to create a custom constructor for the `TransplantAssessmentState` and build shared objects there instead of in this method. At this point the `assess_transplant_request` function is just a constructor for `TransplantAssessmentState` that also calls `check_landing_lints` anyways, so it seems like a reasonable cleanup to take now.

Reply:
I do think that it might be good to at least group the "phids" in their own variable, since it is quite tedious to keep passing multiple variables multiple times. Makes sense to punt refactoring this work to remove most (if not all) the dependency injections into its own PR. We should file a bug for that as well.